### ISSUE TITLE
setItem is damn slow

### ISSUE THREAD

**Issue by lu4** [localForage/localForage#315]
Guys, localForage consumes around 132 seconds to save 10000 objects on my machine. I've looked into implementation, there is a lot of things that can use some optimization. But there's no way to make `setItem` faster, so I would like to add bulk save method (`setItems`) that would work lightning fast, do you have any concerns with it?

Similar to `iterate` functionality, here's a draft interface:

``` javascript
var setItems = function (items, valueFn, keyFn) {
  // If keyFn is not provided, item's collection key will be used as key for the storage
  // Both, objects and arrays of items will be supported
  // Due to performance reasons arrays will be iterated with for (var i = 0; i < arr.length; i++)
  // Objects will be iterated with for (var i in obj) { ... }
  // All of the items will be set in one transaction both for IndexedDB and WebSQL
  // Both keyFn and valueFn will be agnostic to types, the result of a valueFn will get serialized, while the result of keyFn will bet `toString()`'ed
}
```

**Comment by DiegoRBaquero** [comment]
I can't believe this is still an issue after  over a year :/

I'm stuck at the same thing, setItem is damn slow compared to for example localStorage.

+1

**Comment by lu4** [comment]
I've went away with customly-forked implementation, I didn't liked the half-asynchronous / half-synchronous approach, there is no such thing as "half-asynchronous" approach because it's like being half-pregnant, relying on setTimeout is a way to loosing fundamental database guarantee on consistency, when database reports you that everything is saved but it actually isn't. It's a rabbit hole of problems...

**Comment by kaeverens** [comment]
> I can't believe this is still an issue after over a year :/

a year, you say ;-)

**Comment by tofumatt** [comment]
Please note that, for now, there is a setItems plugin: https://github.com/localForage/localForage-setItems <https://github.com/localForage/localForage-setItems> that should help :-)

**Comment by kaeverens** [comment]
> Please note that, for now, there is a setItems plugin: https://github.com/localForage/localForage-setItems <https://github.com/localForage/localForage-setItems> that should help :-)

yeah- I'm testing that at the moment to see if it suits my needs (Cordova app). but because the default databases (indexeddb, localstorage, websql) are not ideal because of their non-guaranteed persistence, I need to combine it with an external plugin for SQLite, which will need hacking to let it understand the non-standard setItems method.

if localForage were to support setItems as an "official" method, then that would encourage third-party plugins to also cater for the method without requiring hacks.

**Comment by tofumatt** [comment]
I think adding those methods to v2 is a great idea. 

- Matt (Sent from mobile)

**Comment by michielbdejong** [comment]
Note that once #656 is merged, the separate `setItems` method becomes unnecessary.

@kaeverens you say you use a custom plugin, so that's a separate story. What speeds are you seeing storing items one-by-one in your external Sqlite database? Is that similarly slow?

**Comment by kaeverens** [comment]
the SQLite library I used was this: https://github.com/thgreasi/localForage-cordovaSQLiteDriver

with this as the Cordova driver: https://github.com/nolanlawson/cordova-plugin-sqlite-2

Before setItems, I was getting a speed of about 10 writes per second on my device (Huawei P20 lite)

After setItems, it's about 150 w/s, so a 15x speedup.
